plotly/presentation_objs/presentation_objs.py

```diff
@@ -0,0 +1,1199 @@
+                    "of the code environment with the '```' characters. For "
+                    "example, your markdown string would include something "
+                    "like:\n\n```python\nx = 2\ny = 1\nprint x\n```\n\n"
+                    "Notice how the language that you want the code to be "
+                    "displayed in is immediately to the right of first "
+                    "entering '```', i.e. '```python'."
+                )
+
+            # find code blocks
+            code_indices = []
+            code_blocks = []
+            wdw_size = len('```')
+            for j in range(len(slide)):
+                if slide[j:j+wdw_size] == '```':
+                    code_indices.append(j)
+
+            for k in range(int(len(code_indices) / 2)):
+                l = 2 * k
+                code_blocks.append(
+                    slide[code_indices[l]:code_indices[l + 1]]
+                )
+
+            lang_and_code_tuples = []
+            for code_block in code_blocks:
+                # validate code blocks
+                code_by_lines = code_block.split('\n')
+                language = _remove_extra_whitespace_from_line(
+                    code_by_lines[0][3:]
+                ).lower()
+                if language == '' or language not in VALID_LANGUAGES:
+                    raise exceptions.PlotlyError(
+                        "The language of your code block should be "
+                        "clearly indicated after the first ``` that "
+                        "begins the code block. The valid languages to "
+                        "choose from are" + list_of_options(
+                            VALID_LANGUAGES
+                        )
+                    )
+                lang_and_code_tuples.append(
+                    (language, '\n'.join(code_by_lines[1:]))
+                )
+
+            # collect text, code and urls
+            title_lines = []
+            url_lines = []
+            text_lines = []
+            inCode = False
+
+            for index, line in enumerate(lines_in_slide):
+                # inCode handling
+                if line[:3] == '```' and len(line) > 3:
+                    inCode = True
+                if line == '```':
+                    inCode = False
+
+                if not inCode and line != '```':
+                    if len(line) > 0 and line[0] == '#':
+                        title_lines.append(line)
+                    elif (_url_parens_contained('Plotly', line) or
+                          _url_parens_contained('Image', line)):
+                        if line.startswith('Plotly(') and 'plot.ly' not in line:
+                            raise exceptions.PlotlyError(
+                                "You are attempting to insert a Plotly Chart "
+                                "in your slide but your url does not have "
+                                "plot.ly in it."
```

Comment:
if the user is using plotly on-premise, then this URL won't necessarily have `plot.ly` in it. It could be like `plotly.acme.com`